Add client properties for secure mds

nsharma-git opened the pull request:
# Description

This PR adds required properties for all CP components to talk to secured Metadata Server.  The detailed doc is [here](https://docs.confluent.io/platform/current/kafka/configure-mds/mds-ssl-config-for-components.html)

Fixes # ([FF-6440](https://confluentinc.atlassian.net/browse/ANSIENG-1024))

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

` molecule --debug converge -s rbac-mds-plain-custom-rhel`
```
PLAY RECAP *********************************************************************
control-center1            : ok=71   changed=30   unreachable=0    failed=0    skipped=40   rescued=0    ignored=0
kafka-broker1              : ok=86   changed=33   unreachable=0    failed=0    skipped=64   rescued=0    ignored=0
kafka-broker2              : ok=79   changed=33   unreachable=0    failed=0    skipped=46   rescued=0    ignored=0
kafka-broker3              : ok=79   changed=33   unreachable=0    failed=0    skipped=46   rescued=0    ignored=0
kafka-connect1             : ok=75   changed=31   unreachable=0    failed=0    skipped=48   rescued=0    ignored=0
kafka-rest1                : ok=71   changed=30   unreachable=0    failed=0    skipped=40   rescued=0    ignored=0
ksql1                      : ok=72   changed=30   unreachable=0    failed=0    skipped=44   rescued=0    ignored=0
mds-zookeeper1             : ok=2    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
schema-registry1           : ok=67   changed=29   unreachable=0    failed=0    skipped=42   rescued=0    ignored=0
zookeeper1                 : ok=61   changed=28   unreachable=0    failed=0    skipped=38   rescued=0    ignored=0

```
MDS runs on secure mode

Verified properties from all components 
Keystore location and path would get added if `ssl_mutual_auth_enabled` is set to true
Kafka broker
```
confluent.metadata.sasl.mechanism=PLAIN
confluent.metadata.security.protocol=SASL_SSL
confluent.metadata.ssl.key.password=xxx
confluent.metadata.ssl.keystore.location=/xxx/kafka_broker.keystore.jks
confluent.metadata.ssl.keystore.password=xxx
confluent.metadata.ssl.truststore.location=/xxx/kafka_broker.truststore.jks
confluent.metadata.ssl.truststore.password=xxx
```
Control Center
```
confluent.metadata.sasl.login.callback.handler.class=io.confluent.kafka.clients.plugins.auth.token.TokenUserLoginCallbackHandler
confluent.metadata.sasl.mechanism=OAUTHBEARER
confluent.metadata.security.protocol=SASL_SSL
confluent.metadata.ssl.truststore.location=/xxxx/control_center.truststore.jks
confluent.metadata.ssl.truststore.password=xxxxxxxxx
```

2. No impact with usecase which doesn't have either `mds_tls_enabled` or `ssl_mutual_auth_enabled` set to true

`molecule --debug converge -s plain-rhel `
```
confluent.metadata.sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username="control_center" password="control_center-secret";
confluent.metadata.sasl.mechanism=PLAIN
confluent.metadata.security.protocol=SASL_PLAINTEXT
```
**Test Configuration**:


# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

JumaX commented on roles/confluent.common/tasks/rbac_setup.yml:
```diff
@@ -18,8 +18,6 @@
   when: (cluster_id_source | default('erp') == 'erp') and not ansible_check_mode
 
-
-# TODO this will not work for Secure ZK or Archive installation method
```
Is this no longer the case?

nsharma-git replied:
The hardcoded path and port has been parameterised in a different PR. The original branch of this PR didn't have this comment which has been removed now. 
So, this task runs with secure zookeeper and archive location now.

JumaX commented on roles/confluent.variables/filter_plugins/filters.py:
```diff
@@ -215,11 +215,6 @@ def client_properties(self, listener_dict, default_ssl_enabled, bouncy_castle_ke
             final_dict[config_prefix + 'ssl.key.password'] = str(keystore_keypass)
 
-        if listener_dict.get('ssl_mutual_auth_enabled', default_ssl_mutual_auth_enabled):
-            final_dict[config_prefix + 'ssl.keystore.location'] = keystore_path
```
Why is this being removed?

nsharma-git replied:
duplicate code

JumaX reviewed (changes requested):
@nsharma-git Can you also confirm the bug number for this?  Field Feedback should be triaged by Product and a bug should be created for us to track our actual work.

nsharma-git commented:
> @nsharma-git Can you also confirm the bug number for this? Field Feedback should be triaged by Product and a bug should be created for us to track our actual work.

Thanks! I have added a corresponding [ANSIENG](https://confluentinc.atlassian.net/browse/ANSIENG-1024) JIRA and updated the in this PR

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -456,10 +456,17 @@ schema_registry_properties:
+                    kerberos_kafka_broker_primary, schema_registry_keytab_path, schema_registry_kerberos_principal|default('kafka'),
+                    false, schema_registry_ldap_user, schema_registry_ldap_password, mds_bootstrap_server_urls) }}"
+  mds_client:
+    enabled: true
```
this is only necessary when rbac is enabled right? I would move it below the rbac props in this dict

same goes for other components

domenicbove replied:
actually based off the jira, it seems the filter is overkill, we only care about ssl props (not all the sasl stuff included in that filter)

nsharma-git replied:
Thanks. Removed the filter and added the properties for secure mds

domenicbove reviewed (changes requested):
looks pretty good, but missing enabled logic

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -456,10 +456,20 @@ schema_registry_properties:
+                    kerberos_kafka_broker_primary, schema_registry_keytab_path, schema_registry_kerberos_principal|default('kafka'),
+                    false, schema_registry_ldap_user, schema_registry_ldap_password, mds_bootstrap_server_urls) }}"
+  mds_client:
+    enabled: rbac_enabled && mds_ssl_enabled
```
```suggestion
    enabled: "{{rbac_enabled and mds_tls_enabled }}"
```
to match formatting elsewhere

nsharma-git replied:
Done

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -456,10 +456,20 @@ schema_registry_properties:
+    properties:
+      confluent.metadata.ssl.truststore.location: "{{schema_registry_truststore_path}}"
+      confluent.metadata.ssl.truststore.password: "{{schema_registry_truststore_storepass}}"
+      confluent.metadata.ssl.keystore.location: "{{schema_registry_keystore_path}}"
+      confluent.metadata.ssl.keystore.password: "{{schema_registry_keystore_storepass}}"
+      confluent.metadata.ssl.key.password: "{{schema_registry_keystore_keypass}}"
+      confluent.metadata.ssl.endpoint.identification.algorithm: HTTPS
```
afaik we do not support mtls on the mds, so truststore should be all you need for one way tls
```suggestion
```

nsharma-git replied:
Done

domenicbove reviewed (changes requested):
i dont think the mds can be mtls w cp-ansible, at least when i was on the team, so there is no need for keystore props

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -1057,17 +1086,25 @@ control_center_properties:
+                    false, sasl_plain_users_final.control_center.principal, sasl_plain_users_final.control_center.password, sasl_scram_users_final.control_center.principal, sasl_scram_users_final.control_center.password, sasl_scram256_users_final.control_center.principal, sasl_scram256_users_final.control_center.password,
+                    kerberos_kafka_broker_primary, control_center_keytab_path, control_center_kerberos_principal|default('c3'),
+                    false, control_center_ldap_user, control_center_ldap_password, mds_bootstrap_server_urls) }}"
+  mds_client:
+    enabled: "{{rbac_enabled and mds_tls_enabled }}"
+    properties:
+      confluent.metadata.ssl.truststore.location: "{{control_center_truststore_path}}"
+      confluent.metadata.ssl.truststore.password: "{{control_center_truststore_storepass}}"
+      confluent.metadata.ssl.endpoint.identification.algorithm: HTTPS
```
can you move each of these blocks down to the rbac blocks, and no need for the newlines

nsharma-git replied:
Moved each block next to "rbac_external_client". Removed extra lines

domenicbove commented on roles/confluent.variables/vars/main.yml:
```diff
@@ -922,10 +944,17 @@ kafka_rest_properties:
+                    kerberos_kafka_broker_primary, kafka_rest_keytab_path, kafka_rest_kerberos_principal|default('rp'),
+                    false, kafka_rest_ldap_user, kafka_rest_ldap_password, mds_bootstrap_server_urls) }}"
+  mds_client:
+    enabled: "{{rbac_enabled and mds_tls_enabled }}"
```
im wondering if these props are needed for rbac **and** centralized rbac or just one of them. can you confirm?

nsharma-git replied:
The reference [document](https://docs.confluent.io/platform/current/kafka/configure-mds/mds-ssl-config-for-components.html) doesn't specifically talk about centralised RBAC. What changes do you expect in case of later
